python/ray/tune/trial_executor.py

```diff
@@ -15,39 +16,75 @@
+# Caching is only helpful/correct for no autoscaler case.
 @lru_cache()
-def _get_warning_threshold() -> float:
+def _get_cluster_resources_no_autoscaler() -> Dict:
+    return ray.cluster_resources()
+
+
+def _can_fulfill_no_autoscaler(trial: Trial) -> bool:
+    """Calculates if there is enough resources for a PENDING trial.
+
+    For no autoscaler case.
+    """
+    assert trial.status == Trial.PENDING
+    total_cpu = trial.resources.cpu + trial.resources.extra_cpu
+    total_gpu = trial.resources.gpu + trial.resources.extra_gpu
+    return total_cpu <= _get_cluster_resources_no_autoscaler().get(
+        "CPU", 0) and total_gpu <= _get_cluster_resources_no_autoscaler().get(
+            "GPU", 0)
+
+
+@lru_cache()
+def _get_insufficient_resources_warning_threshold() -> float:
     if is_ray_cluster():
         return float(
             os.environ.get(
                 "TUNE_WARN_INSUFFICENT_RESOURCE_THRESHOLD_S_AUTOSCALER", "60"))
     else:
+        # Set the default to 10s so that we don't prematurely determine that
+        # a cluster cannot fulfill the resources requirements.
         return float(
-            os.environ.get("TUNE_WARN_INSUFFICENT_RESOURCE_THRESHOLD_S", "1"))
+            os.environ.get("TUNE_WARN_INSUFFICENT_RESOURCE_THRESHOLD_S", "10"))
 
 
 @lru_cache()
-def _get_warning_msg() -> str:
+def _get_insufficient_resources_warning_msg() -> str:
     if is_ray_cluster():
         return (
             f"If autoscaler is still scaling up, ignore this message. No "
             f"trial is running and no new trial has been started within at "
-            f"least the last {_get_warning_threshold()} seconds. "
+            f"least the last {_get_insufficient_resources_warning_threshold()}"
+            f" seconds. "
             f"This could be due to the cluster not having enough "
             f"resources available to start the next trial. Please stop the "
             f"tuning job and readjust resources_per_trial argument passed "
             f"into tune.run() as well as max_workers and worker_nodes "
             f"InstanceType specified in cluster.yaml.")
     else:
         return (f"No trial is running and no new trial has been started within"
-                f" at least the last {_get_warning_threshold()} seconds. "
+                f" at least the last "
+                f"{_get_insufficient_resources_warning_threshold()} seconds. "
                 f"This could be due to the cluster not having enough "
                 f"resources available to start the next trial. Please stop "
                 f"the tuning job and readjust resources_per_trial argument "
                 f"passed into tune.run() and/or start a cluster with more "
                 f"resources.")
 
 
+# A beefed up version when Tune Error is raised.
+def _get_insufficient_resources_error_msg(trial_resources: Resources) -> str:
+    return (f"You ask for {trial_resources.cpu} cpu and {trial_resources.gpu} "
```

Comment:
does this actually work (does it capture the distributed placement group resources too)? @krfricke

Reply:
So the way to test this is just to try using a RLlib agent and make the number of workers > num cpus on the machine. I suspect that `trial_resources.cpu` is not actually the total sum of resources requested. You might need to try something else (like check the placement group and the sum of the resources on the placement group).

Reply:
```suggestion
    return (f"You asked for {trial_resources.cpu} cpu and {trial_resources.gpu} "
```

Reply:
Thanks for the pointer Richard. You are right, for that case, I need to use information from PlacementGroup.
So I used `coin_game_env.py` as my training job. I updated `num_workers` = 17 (I have 16 CPU cores).
On clean master (which currently has the bug and thus spammy):
```
(ray)  xwjiang@xw  ~/ray   master ±  python rllib/examples/coin_game_env.py
2021-09-02 21:13:16,008	INFO services.py:1263 -- View the Ray dashboard at http://127.0.0.1:8266
== Status ==
Memory usage on this node: 37.3/64.0 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/16 CPUs, 0/0 GPUs, 0.0/19.07 GiB heap, 0.0/9.54 GiB objects
Result logdir: /Users/xwjiang/ray_results/PPO_AsymCG
Number of trials: 1/1 (1 PENDING)
+--------------------------+----------+-------+--------+
| Trial name               | status   | loc   |   seed |
|--------------------------+----------+-------+--------|
| PPO_CoinGame_44a68_00000 | PENDING  |       |      0 |
+--------------------------+----------+-------+--------+


2021-09-02 21:13:19,617	WARNING trial_executor.py:247 -- No trial is running and no new trial has been started within at least the last 1.0 seconds. This could be due to the cluster not having enough resources available to start the next trial. Please stop the tuning job and readjust resources_per_trial argument passed into tune.run() and/or start a cluster with more resources.
2021-09-02 21:13:20,618	WARNING trial_executor.py:247 -- No trial is running and no new trial has been started within at least the last 1.0 seconds. This could be due to the cluster not having enough resources available to start the next trial. Please stop the tuning job and readjust resources_per_trial argument passed into tune.run() and/or start a cluster with more resources.
2021-09-02 21:13:21,618	WARNING trial_executor.py:247 -- No trial is running and no new trial has been started within at least the last 1.0 seconds. This could be due to the cluster not having enough resources available to start the next trial. Please stop the tuning job and readjust resources_per_trial argument passed into tune.run() and/or start a cluster with more resources.
2021-09-02 21:13:22,618	WARNING trial_executor.py:247 -- No trial is running and no new trial has been started within at least the last 1.0 seconds. This could be due to the cluster not having enough resources available to start the next trial. Please stop the tuning job and readjust resources_per_trial argument passed into tune.run() and/or start a cluster with more resources.
2021-09-02 21:13:23,619	WARNING trial_executor.py:247 -- No trial is running and no new trial has been started within at least the last 1.0 seconds. This could be due to the cluster not having enough resources available to start the next trial. Please stop the tuning job and readjust resources_per_trial argument passed into tune.run() and/or start a cluster with more resources.
== Status ==
Memory usage on this node: 37.3/64.0 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/16 CPUs, 0/0 GPUs, 0.0/19.07 GiB heap, 0.0/9.54 GiB objects
Result logdir: /Users/xwjiang/ray_results/PPO_AsymCG
Number of trials: 1/1 (1 PENDING)
+--------------------------+----------+-------+--------+
| Trial name               | status   | loc   |   seed |
|--------------------------+----------+-------+--------|
| PPO_CoinGame_44a68_00000 | PENDING  |       |      0 |
+--------------------------+----------+-------+--------+
```

With the change (not spammy as well as raise a TuneError to user with information):
```
(ray)  xwjiang@xw  ~/ray   tmp ±  python rllib/examples/coin_game_env.py
2021-09-02 21:53:20,225	INFO services.py:1263 -- View the Ray dashboard at http://127.0.0.1:8266
== Status ==
Memory usage on this node: 37.3/64.0 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/16 CPUs, 0/0 GPUs, 0.0/18.99 GiB heap, 0.0/9.49 GiB objects
Result logdir: /Users/xwjiang/ray_results/PPO_AsymCG
Number of trials: 1/1 (1 PENDING)
+--------------------------+----------+-------+--------+
| Trial name               | status   | loc   |   seed |
|--------------------------+----------+-------+--------|
| PPO_CoinGame_dd823_00000 | PENDING  |       |      0 |
+--------------------------+----------+-------+--------+


== Status ==
Memory usage on this node: 37.3/64.0 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/16 CPUs, 0/0 GPUs, 0.0/18.99 GiB heap, 0.0/9.49 GiB objects
Result logdir: /Users/xwjiang/ray_results/PPO_AsymCG
Number of trials: 1/1 (1 PENDING)
+--------------------------+----------+-------+--------+
| Trial name               | status   | loc   |   seed |
|--------------------------+----------+-------+--------|
| PPO_CoinGame_dd823_00000 | PENDING  |       |      0 |
+--------------------------+----------+-------+--------+


== Status ==
Memory usage on this node: 37.2/64.0 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/16 CPUs, 0/0 GPUs, 0.0/18.99 GiB heap, 0.0/9.49 GiB objects
Result logdir: /Users/xwjiang/ray_results/PPO_AsymCG
Number of trials: 1/1 (1 PENDING)
+--------------------------+----------+-------+--------+
| Trial name               | status   | loc   |   seed |
|--------------------------+----------+-------+--------|
| PPO_CoinGame_dd823_00000 | PENDING  |       |      0 |
+--------------------------+----------+-------+--------+


Traceback (most recent call last):
  File "rllib/examples/coin_game_env.py", line 83, in <module>
    main(debug_mode, args.stop_iters, args.tf, use_asymmetric_env)
  File "rllib/examples/coin_game_env.py", line 68, in main
    tune_analysis = tune.run(
  File "/Users/xwjiang/ray/python/ray/tune/tune.py", line 585, in run
    runner.step()
  File "/Users/xwjiang/ray/python/ray/tune/trial_runner.py", line 627, in step
    self._run_and_catch(self.trial_executor.on_no_available_trials)
  File "/Users/xwjiang/ray/python/ray/tune/trial_runner.py", line 394, in _run_and_catch
    func(self.get_trials())
  File "/Users/xwjiang/ray/python/ray/tune/trial_executor.py", line 328, in on_no_available_trials
    self._may_warn_insufficient_resources(trials)
  File "/Users/xwjiang/ray/python/ray/tune/trial_executor.py", line 308, in _may_warn_insufficient_resources
    _get_insufficient_resources_error_msg(
  File "/Users/xwjiang/ray/python/ray/tune/trial_executor.py", line 87, in _get_insufficient_resources_error_msg
    trial_cpu_gpu = _get_trial_cpu_and_gpu(trial)
  File "/Users/xwjiang/ray/python/ray/tune/trial_executor.py", line 27, in _get_trial_cpu_and_gpu
    cpu = trial.resources.cpu + trial.resources.extra_cpu
AttributeError: 'Resources' object has no attribute 'resources'
(ray)  ✘ xwjiang@xw  ~/ray   tmp ±  python rllib/examples/coin_game_env.py
2021-09-02 21:54:02,425	INFO services.py:1263 -- View the Ray dashboard at http://127.0.0.1:8266
== Status ==
Memory usage on this node: 37.3/64.0 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/16 CPUs, 0/0 GPUs, 0.0/18.95 GiB heap, 0.0/9.48 GiB objects
Result logdir: /Users/xwjiang/ray_results/PPO_AsymCG
Number of trials: 1/1 (1 PENDING)
+--------------------------+----------+-------+--------+
| Trial name               | status   | loc   |   seed |
|--------------------------+----------+-------+--------|
| PPO_CoinGame_f6a18_00000 | PENDING  |       |      0 |
+--------------------------+----------+-------+--------+


== Status ==
Memory usage on this node: 37.3/64.0 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/16 CPUs, 0/0 GPUs, 0.0/18.95 GiB heap, 0.0/9.48 GiB objects
Result logdir: /Users/xwjiang/ray_results/PPO_AsymCG
Number of trials: 1/1 (1 PENDING)
+--------------------------+----------+-------+--------+
| Trial name               | status   | loc   |   seed |
|--------------------------+----------+-------+--------|
| PPO_CoinGame_f6a18_00000 | PENDING  |       |      0 |
+--------------------------+----------+-------+--------+


== Status ==
Memory usage on this node: 37.3/64.0 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/16 CPUs, 0/0 GPUs, 0.0/18.95 GiB heap, 0.0/9.48 GiB objects
Result logdir: /Users/xwjiang/ray_results/PPO_AsymCG
Number of trials: 1/1 (1 PENDING)
+--------------------------+----------+-------+--------+
| Trial name               | status   | loc   |   seed |
|--------------------------+----------+-------+--------|
| PPO_CoinGame_f6a18_00000 | PENDING  |       |      0 |
+--------------------------+----------+-------+--------+


Traceback (most recent call last):
  File "rllib/examples/coin_game_env.py", line 83, in <module>
    main(debug_mode, args.stop_iters, args.tf, use_asymmetric_env)
  File "rllib/examples/coin_game_env.py", line 68, in main
    tune_analysis = tune.run(
  File "/Users/xwjiang/ray/python/ray/tune/tune.py", line 585, in run
    runner.step()
  File "/Users/xwjiang/ray/python/ray/tune/trial_runner.py", line 627, in step
    self._run_and_catch(self.trial_executor.on_no_available_trials)
  File "/Users/xwjiang/ray/python/ray/tune/trial_runner.py", line 394, in _run_and_catch
    func(self.get_trials())
  File "/Users/xwjiang/ray/python/ray/tune/trial_executor.py", line 328, in on_no_available_trials
    self._may_warn_insufficient_resources(trials)
  File "/Users/xwjiang/ray/python/ray/tune/trial_executor.py", line 307, in _may_warn_insufficient_resources
    raise TuneError(
ray.tune.error.TuneError: You asked for 18.0 cpu and 0.0 gpu per trial, but the cluster only has 16.0 cpu and 0 gpu.The cluster does not have enough resources available to start the next trial. Please stop the tuning job and readjust resources_per_trial argument passed into tune.run() and/or start a cluster with more resources.
```

Richard, so my error information is directing users to `resources_per_trial`, while for this case, we should direct users to `num_workers`.

Reply:
I added both hints in the message. It may look a bit bloated, let me know what you think!

Reply:
Yeah, I really like the new messaging/warning UI.